File: src/leetcode/Javascript/206.Reverse-Linked-list.js
//code solution
const reverseLinkedList = (head) => {
  let previous = null;
  let currentNode = head;
  while (currentNode) {
    // while current node is not null perform operation
    let next = currentNode.next;
    currentNode.next = prev;
    prev = currentNode;
    currentNode = next;
  }
  return previous;
};

//test cases (utility problem)
// 1. Empty Linked List return null >> null
// 2. Single Node  return the node  >> 1-->null
// 3. number additional nodes // 1 --> 2 --> 3 --> 4 --> null  || 4-->3-->2-->1-->null
console.log(reverseLinkedList(empty)); //null
console.log(reverseLinkedList(1)); // 1--> null
console.log(reverseLinkedList([1, 2, 3, 4, 5])) // 5,4,3,2,1

// Complexity Analysis // 5,4,3,2,1 // 5,4,3,2,1
`
    O(n) T - Linear Time - We will iterate n nodes
    O(1) S - Constant Space - No space required
`;

File: src/leetcode/python/206.Reverse LinkedList.py
'''
Given the head of a singly linked list, reverse the list, and return the reversed list.

'''
#Definition of singly linked list
# class Node:
#     def __init__(self, val=0, next= None):
#         self.val = val
#         self.next = next
class Solution:
    #Optimal sln Time Complexity O(n) Space Complexity O(1)
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        #sets the previous node to None
        prev = None
        #sets the current pointer to the head
        current = self.head
        #as long as current is not None
        while current:
            #stores the next node temporarily
            next_node = current.next
            #sets the pointer of the next node to the previous node
            current.next = prev
            #sets the previous node to be our current node
            prev = current
            #updates the current node to be the next node
            current = next_node
        return prev


s = Solution()
print(s.reverseList([1,2,3,4,5])) #Output -> [5,4,3,2,1]
print(s.reverseList([1,2])) # Output -> [2,1]
print(s.reverseList([])) #Output -> []